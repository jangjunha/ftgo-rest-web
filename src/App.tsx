import React, { useEffect, useState } from "react";
import "./App.css";

type FetchState<T> =
  | { stage: "loading" }
  | { stage: "loaded"; value: T }
  | { stage: "error"; message: string };

interface Restaurant {
  id: string;
  name: string;
}

const App = ({ apiEndpoint }: { apiEndpoint: URL }): React.ReactElement => {
  const [restaurants, setRestaurants] = useState<FetchState<Restaurant[]>>({
    stage: "loading",
  });

  useEffect(() => {
    const load = async () => {
      const response = await fetch(new URL("/restaurants/", apiEndpoint));
      if (response.status !== 200) {
        setRestaurants({
          stage: "error",
          message: "Failed to fetch restaurants from server",
        });
        return;
      }

      const res = (await response.json()) as Restaurant[];
      setRestaurants({ stage: "loaded", value: res });
    };
    load();
  });

  return (
    <>
      {restaurants.stage === "loading" && (
        <>
          <p>Loading...</p>
        </>
      )}
      {restaurants.stage === "loaded" && (
        <>
          <p>Restaurants:</p>
          <ul data-testid="restaurant-list">
            {restaurants.value.map((restaurant) => (
              <li key={restaurant.id}>{restaurant.name}</li>
            ))}
          </ul>
        </>
      )}
      {restaurants.stage === "error" && (
        <>
          <p>Error: {restaurants.message}</p>
        </>
      )}
    </>
  );
};

export default App;
