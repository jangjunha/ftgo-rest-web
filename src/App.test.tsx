import path from "path";

import {
  PactV4,
  MatchersV3,
  SpecificationVersion,
} from "@pact-foundation/pact";
import { render, screen } from "@testing-library/react";
import { describe, expect, it } from "vitest";

import App from "./App";

// Create a 'pact' between the two applications in the integration we are testing
const provider = new PactV4({
  dir: path.resolve(process.cwd(), "pacts"),
  consumer: "ftgo-rest-web",
  provider: "ftgo-api-gateway",
  spec: SpecificationVersion.SPECIFICATION_VERSION_V4,
});

describe("GET /restaurants/", () => {
  it("returns an HTTP 200 and a list of restaurants", async () => {
    await provider
      .addInteraction()
      .given("I have a list of restaurants")
      .uponReceiving("a request for all restaurants")
      .withRequest("GET", "/restaurants/")
      .willRespondWith(200, (builder) => {
        builder.headers({ "Content-Type": "application/json" });
        builder.jsonBody(
          MatchersV3.eachLike({
            id: MatchersV3.uuid(),
            name: MatchersV3.string("Nixt Coffee"),
          })
        );
      })
      .executeTest(async (mockserver) => {
        render(<App apiEndpoint={new URL(mockserver.url)} />);

        expect(await screen.findByText("Restaurants:")).toBeInTheDocument();

        const list = await screen.findAllByTestId("restaurant-list");
        expect(list).toMatchSnapshot();
      });
  });
});
