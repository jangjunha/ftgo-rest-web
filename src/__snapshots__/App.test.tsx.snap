// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GET /restaurants/ > returns an HTTP 200 and a list of restaurants 1`] = `
[
  <ul
    data-testid="restaurant-list"
  >
    <li>
      Nixt Coffee
    </li>
  </ul>,
]
`;
